supporting_components/graph.py

```diff
@@ -388,6 +391,48 @@ def is_adjacent(self, u: "Vertex", v: "Vertex") -> bool:
         """
         return v in u.neighbours and (not self.directed or any(e.head == v for e in u.incidence))
 
+    def copy(self):
+        """
+        Returns a copy of the graph.
+        :return: The copy of the graph.
+        """
+        copy = Graph(self.directed)
+        vertices_old_to_new = {}
+
+        for v in self.vertices:
+            vertices_old_to_new[v] = Vertex(copy)
+            vertices_old_to_new[v].label = v.label
+            vertices_old_to_new[v].colornum = v.colornum
+            vertices_old_to_new[v].graph_label = v.graph_label
+
+        for e in self.edges:
+            edge = Edge(vertices_old_to_new[e.tail], vertices_old_to_new[e.head])
+            copy.add_edge(edge)
+
+        return copy
+
+    def is_equal(self, other):
+        """
+        Returns whether or not these two graphs are equal. It checks for three things
+        - If the same vertex is not in the copy.
+        - If the attributes of the vertices with the same label are equal.
+        - If the length of the list of the other graph is empty in the end.
+        This last check should be true, since if a vertex equal to a vertex in the graph is found in the other graph,
+        it is removed from the list of vertices in the other graph.
+        :param other: The other graph to compare this graph with.
+        :return: Whether or not these two graphs are equal.
+        """
+        other_vertices = other.vertices
+        for v in self.vertices:
+            if v in other.vertices:
+                return False
+            for o in other_vertices:
+                if v.label == o.label:
```

Comment:
Deze zou theoretisch in 1 if kunnen denk ik:
```
if v.label == o.label and (v.graph_label != o.graph_label or v.colornum != o.colornum):
```

Reply:
Volgens mij gaat dat niet goed, want ik wil o uit de lijst halen als alle attributen gelijk zijn en False returnen als de labels gelijk zijn, maar niet een van de andere attributen gelijk is. Als ik dit doe `                if v.label == o.label and (v.graph_label != o.graph_label or v.colornum != o.colornum):
                    return False
                else:
                    other_vertices.remove(o)` wat je volgens mij bedoelt, dan haalt hij in de eerste ronde alle elementen o al uit other_vertices en dat wil ik alleen doen voor die o waarvoor dan geldt v.label = o.label

Reply:
Oja, sorry, je hebt gelijk!